geode-common/src/main/java/org/apache/geode/internal/Retry.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.internal;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import org.apache.geode.annotations.VisibleForTesting;
+
+/**
+ * Utility class for retrying operations.
+ */
+public class Retry {
+
+  interface Clock {
+    long nanoTime();
+
+    void sleep(long sleepTime, TimeUnit sleepTimeUnit) throws InterruptedException;
+  }
+
+  private static class SteadyClock implements Clock {
+    @Override
+    public long nanoTime() {
+      return System.nanoTime();
+    }
+
+    @Override
+    public void sleep(long sleepTime, TimeUnit sleepTimeUnit) throws InterruptedException {
```

Comment:
Sleeping doesn't feel like the responsibility of a clock.

Reply:
We can give it another name, like Timer? This is an internal interface made accessible for testing purposes, so I am not too concerned about the naming or break it into another interface.

Reply:
We've been using two separate interfaces in other places in Geode where we do this, and I'd recommend it here too. Lemme go find an example… Ah yes, have a look at `PrimaryHandler` which has `Sleeper` and `MillisecondProvider` interfaces. Hmm looks like `PrimaryHandler.processRequest()` might benefit from this `Retry` class eventually!